gettsim/interface.py

```diff
@@ -105,73 +102,93 @@ def compute_taxes_and_transfers(
+    )
+
     # Load functions.
-    user_functions, internal_functions = load_user_and_internal_functions(functions)
-
-    # Perform several checks on functions and data. Merge internal and user functions.
-    data = copy.deepcopy(data)
-    data = _process_data(data)
-    all_functions = check_data_check_functions_and_merge_functions(
-        user_functions,
-        internal_functions,
-        columns_overriding_functions,
-        targets,
-        data,
-        aggregation_specs,
+    functions_not_overriden, functions_overriden = load_and_check_functions(
+        user_functions_raw=functions,
+        columns_overriding_functions=columns_overriding_functions,
+        targets=targets,
+        data_cols=list(data),
+        aggregation_specs=aggregation_specs,
     )
 
-    # Set up dag.
-    dag = prepare_functions_and_set_up_dag(
-        all_functions=all_functions,
+    # Convert data to expected type.
+    data = _convert_data_to_correct_types(data, functions_overriden)
+
+    # Select necessary nodes by creating a preliminary DAG.
+    nodes = set_up_dag(
+        all_functions=functions_not_overriden,
         targets=targets,
-        params=params,
         columns_overriding_functions=columns_overriding_functions,
         check_minimal_specification=check_minimal_specification,
-        rounding=rounding,
-    )
+    ).nodes
+    necessary_functions = {
+        f_name: f for f_name, f in functions_not_overriden.items() if (f_name in nodes)
+    }
 
-    # Do some checks.
-    _fail_if_root_nodes_are_missing(dag, data)
-    _fail_if_pid_is_non_unique(data)
-    data = _reduce_to_necessary_data(dag, data, check_minimal_specification)
+    # Round and partial all necessary functions.
+    processed_functions = _round_and_partial_parameters_to_functions(
+        necessary_functions, params, rounding
+    )
 
-    # Convert series to numpy arrays
-    data = {key: series.values for key, series in data.items()}
+    # Create input data.
+    input_data = _create_input_data(
+        data=data,
+        processed_functions=processed_functions,
+        targets=targets,
+        columns_overriding_functions=columns_overriding_functions,
+        check_minimal_specification=check_minimal_specification,
+    )
 
-    # Execute DAG.
-    results = execute_dag(dag, data, targets, debug)
+    # Calculate results.
+    tax_transfer_function = dags.concatenate_functions(
+        processed_functions,
+        targets,
+        return_type="dict",
+        aggregator=None,
+        enforce_signature=True,
+    )
+    results = tax_transfer_function(**input_data)
 
     # Prepare results.
-    results = prepare_results(results, data, debug)
+    prepared_results = _prepare_results(results, data, debug)
 
-    return results
+    return prepared_results
 
 
-def check_data_check_functions_and_merge_functions(
-    user_functions,
-    internal_functions,
+def load_and_check_functions(
```

Comment:
I think we want to move this to `functions_loader.py` and probably remove `load_user_and_internal_functions` entirely. Would help with balancing size (1300 / 200 lines) and the division seems rather arbitrary atm.

Unless I am missing something, could you implement that before I continue with the review? Done for the day, anyhow.